# Coding Exercises (Appendix 3)

Now, you will have the opportunity to analyze your own dataset. <br>
__Follow the instructions__ and insert your code! You are either requested to 
- Complete the Code and __Fill in the gaps__. Gaps are marked with "__---__" and are __placeholders__ for your code fragment. 
- Write Code completely __on your own__ 

In some exercises, you will find questions that can only be answered, if your code is correct and returns the right output! The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need a hint, check the __Hints Section__ at the end of this Notebook. Exercises and Hints are numerated accordingly.

If you need some further help or if you want to check your code, you can also watch the __solutions videos__ or check the __solutions notebook__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

### Exercise 9: Pandas - Importing and Inspecting your own Dataset

Before we can work with the Pandas Library, we need to __import Pandas__ (by convention as __pd__). <br>
1. __Import__ Pandas. __Fill in the gaps__!

In [3]:
import pandas as pd

Now we want to import the __Cars Dataset__ that is stored in the CSV-File __cars.csv__. <br>
2. __Import__ the Cars Dataset and __store__ the Pandas DataFrame in the variable __cars__! __Fill in the gaps__!

In [4]:
cars = pd.read_csv('cars.csv')

#### Some more details on the Columns:

- __mpg__: Measures __fuel efficiency__ in terms of miles per gallon. __Higher mpg__ means __higher efficiency__
- __cylinders__: Number of cylinders in the car
- __displacement__: Technical feature (engine)
- __horsepower__: Unit of measurement for engine power 
- __weight__: weight in lbs
- __acceleration__: time to accelerate from 0 to 60 mph (sec.)
- __model_year__: self explanatory
- __origin__: origin of the manufacturer/car
- __name__: model name

3a. __Inspect__ the DataFrame cars by "printing" cars!

In [5]:
cars

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0    18.0          8         307.0       130.0    3504          12.0   
1    15.0          8         350.0       165.0    3693          11.5   
2    18.0          8         318.0       150.0    3436          11.0   
3    16.0          8         304.0       150.0    3433          12.0   
4    17.0          8         302.0       140.0    3449          10.5   
..    ...        ...           ...         ...     ...           ...   
393  27.0          4         140.0        86.0    2790          15.6   
394  44.0          4          97.0        52.0    2130          24.6   
395  32.0          4         135.0        84.0    2295          11.6   
396  28.0          4         120.0        79.0    2625          18.6   
397  31.0          4         119.0        82.0    2720          19.4   

     model_year  origin                       name  
0            70     usa  chevrolet chevelle malibu  
1            70     usa          buick skylark 320  
2            70     usa         plymouth satellite  
3            70     usa              amc rebel sst  
4            70     usa                ford torino  
..          ...     ...                        ...  
393          82     usa            ford mustang gl  
394          82  europe                  vw pickup  
395          82     usa              dodge rampage  
396          82     usa                ford ranger  
397          82     usa             chevrolet s-10  

[398 rows x 9 columns]

3b. __Inspect__ the __first 10 Rows__ of the DataFrame cars! __Fill in the gaps__! What is the __name of the very first car__ in the first row (index label/pos = 0)?

In [7]:
cars.head(10)

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0    3504          12.0   
1  15.0          8         350.0       165.0    3693          11.5   
2  18.0          8         318.0       150.0    3436          11.0   
3  16.0          8         304.0       150.0    3433          12.0   
4  17.0          8         302.0       140.0    3449          10.5   
5  15.0          8         429.0       198.0    4341          10.0   
6  14.0          8         454.0       220.0    4354           9.0   
7  14.0          8         440.0       215.0    4312           8.5   
8  14.0          8         455.0       225.0    4425          10.0   
9  15.0          8         390.0       190.0    3850           8.5   

   model_year origin                       name  
0          70    usa  chevrolet chevelle malibu  
1          70    usa          buick skylark 320  
2          70    usa         plymouth satellite  
3          70    usa              amc rebel sst  
4          70    usa                ford torino  
5          70    usa           ford galaxie 500  
6          70    usa           chevrolet impala  
7          70    usa          plymouth fury iii  
8          70    usa           pontiac catalina  
9          70    usa         amc ambassador dpl

It´s a ... Chevrolet Chevelle Malibu!

3c. Also inspect the __last 5 Rows__! __How many cars/rows__ do we have in the Dataset (mind the range index starting from 0!)?

In [8]:
cars.tail(5)

mpg  cylinders  displacement  horsepower  weight  acceleration  \
393  27.0          4         140.0        86.0    2790          15.6   
394  44.0          4          97.0        52.0    2130          24.6   
395  32.0          4         135.0        84.0    2295          11.6   
396  28.0          4         120.0        79.0    2625          18.6   
397  31.0          4         119.0        82.0    2720          19.4   

     model_year  origin             name  
393          82     usa  ford mustang gl  
394          82  europe        vw pickup  
395          82     usa    dodge rampage  
396          82     usa      ford ranger  
397          82     usa   chevrolet s-10

We have ... 398 Cars!

4. __Inspect__ the entire DataFrame (__all rows__)! __Fill in the gaps!__

In [10]:
pd.options.display.max_rows = 400

In [11]:
cars

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0    18.0          8         307.0       130.0    3504          12.0   
1    15.0          8         350.0       165.0    3693          11.5   
2    18.0          8         318.0       150.0    3436          11.0   
3    16.0          8         304.0       150.0    3433          12.0   
4    17.0          8         302.0       140.0    3449          10.5   
5    15.0          8         429.0       198.0    4341          10.0   
6    14.0          8         454.0       220.0    4354           9.0   
7    14.0          8         440.0       215.0    4312           8.5   
8    14.0          8         455.0       225.0    4425          10.0   
9    15.0          8         390.0       190.0    3850           8.5   
10   15.0          8         383.0       170.0    3563          10.0   
11   14.0          8         340.0       160.0    3609           8.0   
12   15.0          8         400.0       150.0    3761           9.5   
13   14.0          8         455.0       225.0    3086          10.0   
14   24.0          4         113.0        95.0    2372          15.0   
15   22.0          6         198.0        95.0    2833          15.5   
16   18.0          6         199.0        97.0    2774          15.5   
17   21.0          6         200.0        85.0    2587          16.0   
18   27.0          4          97.0        88.0    2130          14.5   
19   26.0          4          97.0        46.0    1835          20.5   
20   25.0          4         110.0        87.0    2672          17.5   
21   24.0          4         107.0        90.0    2430          14.5   
22   25.0          4         104.0        95.0    2375          17.5   
23   26.0          4         121.0       113.0    2234          12.5   
24   21.0          6         199.0        90.0    2648          15.0   
25   10.0          8         360.0       215.0    4615          14.0   
26   10.0          8         307.0       200.0    4376          15.0   
27   11.0          8         318.0       210.0    4382          13.5   
28    9.0          8         304.0       193.0    4732          18.5   
29   27.0          4          97.0        88.0    2130          14.5   
30   28.0          4         140.0        90.0    2264          15.5   
31   25.0          4         113.0        95.0    2228          14.0   
32   25.0          4          98.0         NaN    2046          19.0   
33   19.0          6         232.0       100.0    2634          13.0   
34   16.0          6         225.0       105.0    3439          15.5   
35   17.0          6         250.0       100.0    3329          15.5   
36   19.0          6         250.0        88.0    3302          15.5   
37   18.0          6         232.0       100.0    3288          15.5   
38   14.0          8         350.0       165.0    4209          12.0   
39   14.0          8         400.0       175.0    4464          11.5   
40   14.0          8         351.0       153.0    4154          13.5   
41   14.0          8         318.0       150.0    4096          13.0   
42   12.0          8         383.0       180.0    4955          11.5   
43   13.0          8         400.0       170.0    4746          12.0   
44   13.0          8         400.0       175.0    5140          12.0   
45   18.0          6         258.0       110.0    2962          13.5   
46   22.0          4         140.0        72.0    2408          19.0   
47   19.0          6         250.0       100.0    3282          15.0   
48   18.0          6         250.0        88.0    3139          14.5   
49   23.0          4         122.0        86.0    2220          14.0   
50   28.0          4         116.0        90.0    2123          14.0   
51   30.0          4          79.0        70.0    2074          19.5   
52   30.0          4          88.0        76.0    2065          14.5   
53   31.0          4          71.0        65.0    1773          19.0   
54   35.0          4          72.0        69.0

5. Get some __meta information__ on our DataFrame! In which __Column__ do we have __missing/NaN__ Values?

In [12]:
cars.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        398 non-null    int64  
 5   acceleration  398 non-null    float64
 6   model_year    398 non-null    int64  
 7   origin        398 non-null    object 
 8   name          398 non-null    object 
dtypes: float64(4), int64(3), object(2)
memory usage: 28.1+ KB


We have 6 missing/NaN Values in the column ... horsepower!

6. Let´s get some __summary statistics__ on our DataFrame! What is the __maximum__ value in the column __horsepower__?

In [13]:
cars.describe()

mpg   cylinders  displacement  horsepower       weight  \
count  398.000000  398.000000    398.000000  392.000000   398.000000   
mean    23.514573    5.454774    193.425879  104.469388  2970.424623   
std      7.815984    1.701004    104.269838   38.491160   846.841774   
min      9.000000    3.000000     68.000000   46.000000  1613.000000   
25%     17.500000    4.000000    104.250000   75.000000  2223.750000   
50%     23.000000    4.000000    148.500000   93.500000  2803.500000   
75%     29.000000    8.000000    262.000000  126.000000  3608.000000   
max     46.600000    8.000000    455.000000  230.000000  5140.000000   

       acceleration  model_year  
count    398.000000  398.000000  
mean      15.568090   76.010050  
std        2.757689    3.697627  
min        8.000000   70.000000  
25%       13.825000   73.000000  
50%       15.500000   76.000000  
75%       17.175000   79.000000  
max       24.800000   82.000000

The maximum value in the column horsepower is ... 230!

-------------------------------

# Well Done!

---------------------------------------------------------------------------------------------------------------------

# Hints (Spoiler!)

1. pandas as pd

2. Pass the filename in quotation marks to pd.read_csv()

3a. cars

3b. head() method with n = 10

3c. tail() method

4. max_rows 

5. info() method

6. describe() method